core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h

```diff
@@ -197,6 +197,25 @@ namespace ledger {
                                account->getKeychain()->getRestoreKey(),
                                account->getWallet()->getName(), DateUtils::toJSON(buddy->startDate));
 
+		// extend the keychain to cover all the addresses
+                if (buddy->wallet->getWalletType() == api::WalletType::BITCOIN)
+                {
+                    int currentBatchIndex = 0;
+                    int txSize;
+                    do {
+                        auto batch = vector::map<std::string, std::shared_ptr<AddressType>>(
+                            buddy->keychain->getAllObservableAddresses(currentBatchIndex * buddy->halfBatchSize, (currentBatchIndex + 1) * buddy->halfBatchSize - 1),
+                            [](const std::shared_ptr<AddressType>& addr) -> std::string {
+                                return addr->toString();
+                            }
+                        );
+                        auto txs = ledger::core::async::wait(_explorer->getTransactions(batch, optional<std::string>(), optional<void*>()));
```

Comment:
We must not wait, asynchronous task, waiting on a thread mean using 2 thread just for one task. This part needs to be rewritten, to avoid locking. You can use a combination of Future and Future::flatMap to avoid locking threads.

Reply:
@pollastri-pierre Fixed. Could you please review my change again? Thanks